File: src/main/java/fr/adrienbrault/idea/symfony2plugin/ux/variable/collector/UxComponentVariableCollector.java
package fr.adrienbrault.idea.symfony2plugin.ux.variable.collector;

import com.intellij.openapi.project.Project;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.psi.elements.Field;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import fr.adrienbrault.idea.symfony2plugin.templating.variable.TwigFileVariableCollector;
import fr.adrienbrault.idea.symfony2plugin.templating.variable.TwigFileVariableCollectorParameter;
import fr.adrienbrault.idea.symfony2plugin.templating.variable.dict.PsiVariable;
import fr.adrienbrault.idea.symfony2plugin.util.UxUtil;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.HashSet;
import java.util.Map;

/**
 * Support variables piped from "AsTwigComponent, AsLiveComponent" to Twig templates
 *
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class UxComponentVariableCollector implements TwigFileVariableCollector {
    @Override
    public void collectPsiVariables(@NotNull TwigFileVariableCollectorParameter parameter, @NotNull Map<String, PsiVariable> variables) {
        Project project = parameter.getProject();
        Collection<PhpClass> componentClassesForTemplateFile = UxUtil.getComponentClassesForTemplateFile(project, parameter.getElement().getContainingFile());

        if (!componentClassesForTemplateFile.isEmpty()) {
            variables.put("attributes", new PsiVariable("\\Symfony\\UX\\TwigComponent\\ComponentAttributes"));
        }

        for (PhpClass phpClass : componentClassesForTemplateFile) {
            variables.put("this", new PsiVariable(phpClass.getFQN()));

            for (Field field : phpClass.getFields()) {
                if (field.getModifier().isPublic()) {
                    variables.put(field.getName(), new PsiVariable(
                        PhpIndex.getInstance(project).completeType(project, field.getType(), new HashSet<>()).getTypes()
                    ));
                }
            }
        }
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/ux/variable/collector/UxComponentVariableCollector.kt
package fr.adrienbrault.idea.symfony2plugin.ux.variable.collector

import com.jetbrains.php.PhpIndex
import fr.adrienbrault.idea.symfony2plugin.templating.variable.TwigFileVariableCollector
import fr.adrienbrault.idea.symfony2plugin.templating.variable.TwigFileVariableCollectorParameter
import fr.adrienbrault.idea.symfony2plugin.templating.variable.dict.PsiVariable
import fr.adrienbrault.idea.symfony2plugin.util.UxUtil

/**
 * Support variables piped from "AsTwigComponent, AsLiveComponent" to Twig templates
 *
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class UxComponentVariableCollector : TwigFileVariableCollector {
    override fun collectPsiVariables(parameter: TwigFileVariableCollectorParameter, variables: MutableMap<String, PsiVariable>) {
        val project = parameter.project
        val componentClassesForTemplateFile = UxUtil.getComponentClassesForTemplateFile(project, parameter.element.containingFile)

        if (componentClassesForTemplateFile.isNotEmpty()) {
            variables["attributes"] = PsiVariable("\\Symfony\\UX\\TwigComponent\\ComponentAttributes")
        }

        for (phpClass in componentClassesForTemplateFile) {
            variables["this"] = PsiVariable(phpClass.fqn)

            for (field in phpClass.fields) {
                if (field.modifier.isPublic) {
                    variables[field.name] = PsiVariable(
                        PhpIndex.getInstance(project).completeType(project, field.type, HashSet()).types
                    )
                }
            }
        }
    }
}
